nfs: nfs provisioner controlled by operator

prksu opened the pull request:
Signed-off-by: Ahmad Nurus S <prksu.sh@gmail.com>

<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/master/development-flow.html)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->

**Description of your changes:**

Implementation of nfs provisioner controlled by the operator.
design docs https://github.com/rook/rook/pull/5830

**Which issue is resolved by this Pull Request:**
Resolves #4982 and Resolves #3074 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[test nfs]

ashishranjan738 commented on pkg/operator/nfs/controller.go:
```diff
@@ -158,8 +167,8 @@ func (r *NFSServerReconciler) reconcileNFSServerConfig(ctx context.Context, cr *
 		nfsGaneshaConfig := `
 EXPORT {
 	Export_Id = ` + fmt.Sprintf("%v", id) + `;
-	Path = /` + claimName + `;
```
how about making export as a constant, since it is getting used at multiple places.

ashishranjan738 commented on pkg/operator/nfs/controller.go:
```diff
@@ -287,7 +296,7 @@ func (r *NFSServerReconciler) reconcileNFSServer(ctx context.Context, cr *nfsv1a
 
 			volumeMounts = append(volumeMounts, corev1.VolumeMount{
 				Name:      shareName,
-				MountPath: "/" + claimName,
```
same here

ashishranjan738 commented on pkg/operator/nfs/controller.go:
```diff
@@ -313,3 +322,66 @@ func (r *NFSServerReconciler) reconcileNFSServer(ctx context.Context, cr *nfsv1a
+				VolumeSource: v1.VolumeSource{
+					NFS: &v1.NFSVolumeSource{
+						Server: nfsserversvc.Spec.ClusterIP,
+						Path:   "/export/" + claimName,
```
same here

ashishranjan738 commented on pkg/operator/nfs/controller.go:
```diff
@@ -313,3 +322,66 @@ func (r *NFSServerReconciler) reconcileNFSServer(ctx context.Context, cr *nfsv1a
+
+			volumeMounts = append(volumeMounts, v1.VolumeMount{
+				Name:      shareName,
+				MountPath: filepath.Join("/export/", claimName),
```
same here

ashishranjan738 commented on pkg/operator/nfs/provisioner.go:
```diff
@@ -18,62 +18,89 @@ package nfs
 
-	if options.PVC.Spec.Selector != nil {
-		return nil, fmt.Errorf("claim Selector is not supported")
+	serverName, present := sc.Parameters["nfsServerName"]
```
can we make these map keys as constant

ashishranjan738 commented on pkg/operator/nfs/provisioner.go:
```diff
@@ -86,67 +113,78 @@ func (p *nfsProvisioner) Provision(options controller.ProvisionOptions) (*v1.Per
+	var exportPath string
+	for _, export := range nfsserver.Spec.Exports {
+		if export.Name == exportName {
+			exportPath = filepath.Join(mountPath, export.PersistentVolumeClaim.ClaimName)
```
how about logging something if exportName is not matched, that would be helpful for debugging

rohan47 commented on pkg/operator/nfs/provisioner.go:
```diff
@@ -86,67 +113,78 @@ func (p *nfsProvisioner) Provision(options controller.ProvisionOptions) (*v1.Per
 
-func (p *nfsProvisioner) Delete(volume *v1.PersistentVolume) error {
-	return nil
+func (p *Provisioner) Delete(volume *v1.PersistentVolume) error {
```
Can we have unit test for this function.

rohan47 commented on pkg/operator/nfs/controller.go:
```diff
@@ -313,3 +322,66 @@ func (r *NFSServerReconciler) reconcileNFSServer(ctx context.Context, cr *nfsv1a
 	}
 }
+
+func (r *NFSServerReconciler) reconcileNFSProvisioner(ctx context.Context, cr *nfsv1alpha1.NFSServer) error {
```
Unit test for this

prksu replied:
the current unit test already covered this function

rohan47 commented on pkg/operator/nfs/controller_test.go:
```diff
@@ -147,6 +149,48 @@ func TestNFSServerReconciler_Reconcile(t *testing.T) {
+		return svc
+	}
+
+	expectedProvisioner := func(scheme *runtime.Scheme, cr *nfsv1alpha1.NFSServer) *appsv1.Deployment {
```
Update the unit test to verify this

prksu replied:
updated, thanks

prksu commented:
This PR also fixes https://github.com/rook/rook/issues/3074

prksu commented:
@rohan47 @ashishranjan738 I've added unit test for nfs provisioner and also updated some constant as per as your review. please take another look. thanks!

prksu commented:
@rohan47 @ashishranjan738 i've updated this PR to make nfs-provisioner as nfs-server sidecar. please take another look. thanks
fyi @jbw976

ashishranjan738 commented:
looks good to me. @prksu can you fix the ci errors

ashishranjan738 commented:
looks good to me.
